### DISCUSSION TITLE
m_iMatchStats_Damage

### DISCUSSION THREAD

**Discussion by mavrick** [saul/demofile#220] [Questions] [answered]
Hey mate,

Happy new years! 

I'm having a bit of trouble trying to get the damage dealt per user per round with ``m_iMatchStats_Damage`` 

I've tried using:

```javascript
let pr = this._demo.entities.getSingleton('DT_CSPlayer');
let values = pr.props['m_iMatchStats_Damage'];
return values[Object.keys(values)[this.index]];	
```

But it just results with a zero ``0`` value.

Ideally I want to calculate damage dealt per round and average damage per player.

**Comment by mavrick** [comment]
```javascript
get damage() {
  let clientSlot = this.clientSlot;
  if(clientSlot.toString().length === 1) {
    clientSlot = `00${clientSlot}`;
  } else if(clientSlot.toString().length === 2) {
    clientSlot = `0${clientSlot}`;
  }
  return this.props['m_iMatchStats_Damage'][clientSlot];
}
```
doesn't feel right either. is this total damage per player over the time of the entire map?

**Comment by saul** [comment]
### Edit: this post is incorrect, m_iMatchStats do not represent inter-player damage statistics.

------

Hi mavrick

You want to do something like:

```js
damage(victim) {
  let clientSlot = victim.clientSlot;
  if(clientSlot.toString().length === 1) {
    clientSlot = `00${clientSlot}`;
  } else if(clientSlot.toString().length === 2) {
    clientSlot = `0${clientSlot}`;
  }
  return this.getProp('m_iMatchStats_Damage', clientSlot);
}
```

Note how you need to use the client slot of the *victim*, otherwise it's most likely that a player won't really damage themselves during the course of the match.

Also, judging by the name 'match stats', I assume that it is per match as opposed to per round.

Thanks

**Comment by mavrick** [comment]
Yea right. So we need to calculate the damage dealt by a player to each other player to determine total damage dealt per player. And as each round_end is triggered we can calculate the difference to determine round by round damage per player.

**Comment by saul** [comment]
That sounds right, yes

**Comment by Brammz** [comment]
Hi,

I am trying to do the same thing, but when I execute this code most numbers stay at 0. I'm not sure whether I'm using the code correctly (my JS knowledge is less advanced). I am executing the following code:

```
demoFile.gameEvents.on('round_end', e => {
  for (p of teamT.members) {
    let clientSlot = p.clientSlot;
    if(clientSlot.toString().length === 1) {
      clientSlot = `00${clientSlot}`;
    } else if(clientSlot.toString().length === 2) {
      clientSlot = `0${clientSlot}`;
    }
    console.log(p.name + ': ' + p.getProp('m_iMatchStats_Damage', clientSlot));
  }
});
```

Also, is this the only way to do it?

**Comment by saul** [comment]
@Brammz / @mavrick:

It turns out that `m_iMatchStats_*` do not represent how much damage players do to each other during the match. They represent the total damage each player does during a round. I've just released demofile 0.4.13 which adds `Player#matchStats`:

```js
  /**
   * @returns {Object[]} Object representing the player's performance per round
   * @property {string} kills - Kills dealt
   * @property {string} damage - Damage dealt
   * @property {string} equipmentValue - Total value of equipment (snapshot at the end of buy time)
   * @property {string} moneySaved - Money remaining (snapshot at the end of buy time)
   * @property {string} killReward - Money earnt by killing other players
   * @property {string} liveTime - Number of seconds this player stayed alive (zero if survived the whole round)
   * @property {string} deaths - Number of times the player died this round
   * @property {string} assists - Kills assisted by this player
   * @property {string} headShotKills - Kills that were headshots
   * @property {string} objective - Number of times the player achieved the objective (e.g defused the bomb or planted the exploding bomb)
   */
```

**Comment by Brammz** [comment]
Oh ok, thank you!
I'm now trying the following code but it doesn't seem to output the correct values (also happens for .kills, .deaths, ...).

```
demoFile.gameEvents.on('round_end', e => {
    let roundNumber = demoFile.gameRules.roundsPlayed;
    for (p of teamT.members) {
        console.log(p.name + ':' + p.matchStats[roundNumber].damage);
    }
});
```

**Comment by saul** [comment, marked as answer]
You needs to subtract 1 from rounds played

**Comment by Brammz** [comment]
This does indeed work for damage dealt. However, it is not correct for kills/assists/deaths.
Also, damage dealt to armor is not included in the damage dealt. Is there any way to get this info?

**Comment by saul** [comment]
Kills/assists/deaths looked fine to me. Do you have a demo that exhibits this issue?

Not sure about armor, you may have to manually track that in player_hurt

**Comment by Brammz** [comment]
This is the demo that I am currently using: https://www.hltv.org/matches/2317297/nip-vs-faze-iem-oakland-2017. I am getting rounds with 11 kills on one player. `Player#kills` does give me the correct values.

I am using this parser for my master's thesis and have some more questions about it, is there a way I can contact you more easily?

**Comment by mavrick** [comment]
ensure you are respecting ``begin_new_match`` and reset your stats because the players can get frags in warmup mode that will be tracked under ``player_death``

for player damage use ``player_hurt``
